src/main/java/com/google/devtools/build/lib/rules/proto/ProtoToolchainInfo.java

```diff
@@ -0,0 +1,44 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.rules.proto;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.devtools.build.lib.analysis.FilesToRunProvider;
+import com.google.devtools.build.lib.analysis.platform.ToolchainInfo;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import com.google.devtools.build.lib.skyframe.serialization.autocodec.AutoCodec;
+import com.google.devtools.build.lib.starlarkbuildapi.proto.ProtoToolchainInfoApi;
+
+/**
+ * Information about the tools used by the <code>proto_*</code> and <code>LANG_proto_*</code> rules.
+ */
+@Immutable
+@AutoCodec
+public class ProtoToolchainInfo extends ToolchainInfo implements ProtoToolchainInfoApi<FilesToRunProvider> {
```

Comment:
What we've learned is that extending ToolchainInfo isn't a good idea: Starlark doesn't have the concept of inheritance and so using it for native providers leads to a mismatch.

Instead, this class should just be "ProtoToolchainInfo extends NativeInfo", and in the actual ProtoToolchain rule implementation, create a new ToolchainInfo that contains the ProtoToolchainInfo:
```
ProtoToolchainInfo protoToolchainInfo = new ProtoToolchainInfo(...);
ToolchainInfo toolchainInfo = new ToolchainInfo(ImmutableMap.of("protoToolchain", protoToolchainInf);

...

ruleBuilder
    .addProvider(protoToolchainInfo)
    .addProvider(toolchainInfo)
    .build()
```

Then, rules that require the toolchain can get the ToolchainInfo and find the ProtoToolchainInfo by calling:
```
ToolchainInfo toolchainInfo = ruleContext.getToolchainContext().forToolchainType(toolchainType);
ProtoToolchainInfo protoToolchainInfo = (ProtoToolchainInfo) toolchainInfo.get("protoToolchain");
```

(I'm translating this from the Starlark version, so my syntax may not be 100% correct).